src/renderer/wg_engine/tvgWgRenderer.cpp

```diff
@@ -95,10 +237,206 @@ bool WgRenderer::clear() {
+            renderPassDesc.occlusionQuerySet = nullptr;
+            renderPassDesc.timestampWriteCount = 0;
+            renderPassDesc.timestampWrites = nullptr;
+            // begin render pass
+            WGPURenderPassEncoder renderPassEncoder = wgpuCommandEncoderBeginRenderPass(commandEncoder, &renderPassDesc); {
+                // iterate render data
+                for (size_t i = 0; i < mRenderDatas.count; i++) {
+                    // get render data
+                    WgRenderDataShape* renderData = (WgRenderDataShape*)(mRenderDatas[i]);
+                    
+                    // iterate geometry data
+                    for (uint32_t j = 0; j < renderData->mGeometryDataFill.count; j++) {
+                        // draw to stencil
+                        mPipelineEmpty.set(renderPassEncoder);
+                        mPipelineBindGroupEmpty.bind(renderPassEncoder, 0);
+                        renderData->mGeometryDataFill[j]->draw(renderPassEncoder);
+
+                        // draw brush
+                        renderData->mPipelineBase->set(renderPassEncoder);
+                        renderData->mPipelineBindGroup->bind(renderPassEncoder, 0);
+                        mGeometryDataPipeline.draw(renderPassEncoder);
+                    }
+                }
+            }
+            // end render pass
+            wgpuRenderPassEncoderEnd(renderPassEncoder);
+            wgpuRenderPassEncoderRelease(renderPassEncoder);
+
+            // release backbuffer and present
+            wgpuTextureViewRelease(backBufferView);
+        }
+        commandsBuffer = wgpuCommandEncoderFinish(commandEncoder, &commandBufferDesc);
+        wgpuCommandEncoderRelease(commandEncoder);
+    }
+    // queue submit and command buffer release
+    wgpuQueueSubmit(mQueue, 1, &commandsBuffer);
+    wgpuCommandBufferRelease(commandsBuffer);
+    
+    // go to the next frame
+    wgpuSwapChainPresent(mSwapChain);
+
+    // clear current render datas
+    mRenderDatas.clear();
     return true;
 }
 
+bool WgRenderer::target(uint32_t* buffer, uint32_t stride, uint32_t w, uint32_t h) {
+    // check properties
+    assert(w > 0);
+    assert(h > 0);
+
+    // store target surface properties
+    mTargetSurface.stride = stride;
+    mTargetSurface.w = w;
+    mTargetSurface.h = h;
+
+    // update view matrix
+    mViewMatrix[ 0] = +2.0f / w; mViewMatrix[ 1] = +0.0f;     mViewMatrix[ 2] = +0.0f; mViewMatrix[ 3] = +0.0f;
+    mViewMatrix[ 4] = +0.0f;     mViewMatrix[ 5] = -2.0f / h; mViewMatrix[ 6] = +0.0f; mViewMatrix[ 7] = +0.0f;
+    mViewMatrix[ 8] = +0.0f;     mViewMatrix[ 9] = +0.0f;     mViewMatrix[10] = -1.0f; mViewMatrix[11] = +0.0f;
+    mViewMatrix[12] = -1.0f;     mViewMatrix[13] = +1.0f;     mViewMatrix[14] = +0.0f; mViewMatrix[15] = +1.0f;
+
+    // return result
+    return true;
+}
+
+// target for native window handle
 bool WgRenderer::target(void* window, uint32_t w, uint32_t h) {
+    // check properties
+    assert(window);
```

Comment:
rather than adding assert() here, please add exception handling in WgCanvas layer and return the proper values(tvg::Result)